### PR TITLE
fix(diagnostic): always return copies of diagnostic items

### PR DESCRIPTION
**Pull request by echasnovski**

### CODE REVIEW

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -756,7 +756,7 @@ end
 ---                        - namespace: (number) Limit diagnostics to the given namespace.
 ---                        - lnum: (number) Limit diagnostics to the given line number.
 ---                        - severity: See |diagnostic-severity|.
----@return Diagnostic[] table A list of diagnostic items |diagnostic-structure|. Keys `bufnr`, `end_lnum`, `end_col`, and `severity` are guaranteed to be present.
+---@return Diagnostic[] table A list of copies of diagnostic items |diagnostic-structure|. Keys `bufnr`, `end_lnum`, `end_col`, and `severity` are guaranteed to be present.
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:759]: Let's leave this as is and add a note to the general function description.

```suggestion
---@return Diagnostic[] table A list of diagnostic items |diagnostic-structure|. Keys `bufnr`, `end_lnum`, `end_col`, and `severity` are guaranteed to be present.
```

Comment (gpanders) [reply runtime/lua/vim/diagnostic.lua:759]: I can't leave a comment on line 751 (pls fix GitHub), but suggestion (modulo line wrapping):

```
--- Get current diagnostics.
---
--- Modifying diagnostics in the returned table has no effect. To set diagnostics in a buffer, use |vim.diagnostic.set()|.
```

**Comment by echasnovski** [comment]
Thanks, Gregory!

**Comment by lewis6991** [comment]
Someone in chat noticed this causing perf issues when showing diagnostics in the statusline.

Is there something we can do to resolve this? I could see why using `deepcopy()` on a lot of diagnostics would be an issue.

**Comment by lewis6991** [comment]
Someone in chat noticed this causing perf issues when showing diagnostics in the statusline.

Is there something we can do to resolve this? I could see why using `deepcopy()` on a lot of diagnostics would be an issue.

**Comment by echasnovski** [comment]
> Someone in chat noticed this causing perf issues when showing diagnostics in the statusline.
> 
> Is there something we can do to resolve this? I could see why using `deepcopy()` on a lot of diagnostics would be an issue.

I think using `deepcopy()` is necessary evil to ensure that those can not be overridden in place by not original diagnostics provider.

Making a dedicated function designed to be used in statusline without much overhead seems like a good idea to me. Previously there was a `vim.lsp.diagnostic.get_count()` which is now deprecated in favor of `vim.diagnostic.get()`. Maybe a new `vim.diagnostic.get_info()` which at least would return something like `{ count = { [vim.diagnostic.severity.ERROR] = <number>, ... } }`?

File: runtime/doc/diagnostic.txt [hunk 2]
```diff
@@ -480,6 +480,9 @@ fromqflist({list})                               *vim.diagnostic.fromqflist()*
 get({bufnr}, {opts})                                    *vim.diagnostic.get()*
     Get current diagnostics.
 
+    Modifying diagnostics in the returned table has no effect. To set
```
Comment (shahardra) [inline runtime/doc/diagnostic.txt:483]: Shh

**Comment by gpanders** [comment]
The removal of `vim.lsp.diagnostic.get_count()` was intentional, but now that `vim.diagnostic.get()` has become quite heavy it might be worthwhile to add `vim.diagnostic.count()` that has a fast path and avoids copying anything.

**Comment by lewis6991** [comment]
#26858 should alleviate things a bit with `deepcopy` by at least 2x.

**Pull request merged**